include/circt/Dialect/ESI/ESIServices.td

```diff
@@ -0,0 +1,134 @@
+//===- ESIServices.td - ESI services ops -------------------*- tablegen -*-===//
+//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+include "circt/Dialect/HW/HWTypes.td"
+include "mlir/IR/RegionKindInterface.td"
+
+def ServiceDeclOp : ESI_Op<"service.decl",
+          [SingleBlock, NoTerminator, HasParent<"::mlir::ModuleOp">, Symbol]> {
+  let summary = "An ESI service interface declaration";
+  let description = [{
+    A declaration of an ESI service interface. Defines a contract between a
+    service provider and its clients.
+
+    Example:
+
+    ```mlir
+    esi.service.decl @HostComms {
+      esi.service.to_server send : !esi.channel<!esi.any>
+      esi.service.to_client recieve : !esi.channel<i8>
+    }
+    ```
+  }];
+
+  let arguments = (ins SymbolNameAttr:$sym_name);
+  let regions = (region SizedRegion<1>:$ports);
+
+  let assemblyFormat = [{
+    $sym_name $ports attr-dict
+  }];
+}
+
+def ToServerOp : ESI_Op<"service.to_server",
+                        [HasParent<"::circt::esi::ServiceDeclOp">]> {
+  let summary = "An ESI service port headed to the service";
+
+  let arguments = (ins SymbolNameAttr:$inner_sym, TypeAttr:$type);
+  let assemblyFormat = [{
+    $inner_sym attr-dict `:` $type
+  }];
+}
+
+def ToClientOp : ESI_Op<"service.to_client",
+                        [HasParent<"::circt::esi::ServiceDeclOp">]> {
+  let summary = "An ESI service port headed to a particular client";
+
+  let arguments = (ins SymbolNameAttr:$inner_sym, TypeAttr:$type);
+  let assemblyFormat = [{
+    $inner_sym  attr-dict `:` $type
+  }];
+}
+
+def ServiceInstanceOp : ESI_Op<"service.instance"> {
+  let summary = "Instantiate a server module";
+  let description = [{
+    Instantiate a service adhering to a service declaration interface.
+
+    A pass collects all of the connection requests to the service this op
+    implements from the containing modules' descendants (in the instance
+    hierarchy). It bubbles them all up to the module containing this op,
+    creating the necessary ESI channel ports, groups them appropriately, then
+    replaces this op with a `service.impl_req`.
+
+    Since implementing the server will usually need "normal" I/O, `inputs` and
+    `results` act like normal `hw.instance` ports.
+
+    $identifier is used by frontends to specify or remember the type of
+    implementation to use for this service.
```

Comment:
I think i understand the intention here, but i find the IR a bit difficult to reason about (`%c = esi.service.instance "topComms" @HostComms (%clk) : (i1) -> (i8)`) in terms of what's doing what between `"topComms"` and `@HostComms`. Could the op format be modified to better reflect the intention of the attribute?
e.g.:
```mlir
%c = esi.service.instance @HostComms impl as "topComms" (%clk) : (i1) -> (i8)
```